File: src/aws/lambdas/incremental_distribution/src/main/java/uk/nhs/nhsx/analyticsevents/PayloadValidator.java
package uk.nhs.nhsx.analyticsevents;

import com.fasterxml.jackson.core.type.TypeReference;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import uk.nhs.nhsx.core.Jackson;

import java.io.IOException;
import java.util.List;
import java.util.Map;
import java.util.Optional;

public class PayloadValidator {

    private final static Logger log = LogManager.getLogger(PayloadValidator.class);

    public Optional<Map<String, Object>> maybeValidPayload(String requestBody) {
        try {
            Map<String, Object> payload = Jackson.readJson(requestBody, new TypeReference<>() {
            });

            validateMetadata(payload);
            validateEvents(payload);

            return Optional.of(payload);
        } catch (IOException | ValidationException | IllegalArgumentException e) {
            log.info("Payload was invalid", e);
            return Optional.empty();
        }
    }

    private void validateMetadata(Map<String, ?> payload) throws ValidationException {
        requireMapAttribute(payload, "metadata");

        Object metadataRaw = payload.get("metadata");
        if (!(metadataRaw instanceof Map)) {
            throw new ValidationException("metadata must be a map");
        }

        Map<?, ?> metadata = (Map<?, ?>) metadataRaw;

        requireMapAttribute(metadata, "operatingSystemVersion");
        requireMapAttribute(metadata, "latestApplicationVersion");
        requireMapAttribute(metadata, "deviceModel");
        requireMapAttribute(metadata, "postalDistrict");
    }

    private void validateEvents(Map<String, ?> payload) throws ValidationException {
        requireMapAttribute(payload, "events");

        Object eventsRaw = payload.get("events");

        if (!(eventsRaw instanceof List)) {
            throw new ValidationException("events must be a list");
        }

        List<?> events = (List<?>) eventsRaw;

        for (Object event : events) {
            requireMapAttribute(event, "type");
            requireMapAttribute(event, "version");
            requireMapAttribute(event, "payload");
        }
    }

    private void requireMapAttribute(Object mapRaw, String key) throws ValidationException {
        if (!(mapRaw instanceof Map)) {
            throw new ValidationException("expected map type but was " + mapRaw.getClass().getSimpleName());
        }

        Map<?, ?> map = (Map<?, ?>) mapRaw;

        if (!(map.containsKey(key))) {
            throw new ValidationException("Map did not contain expected key + " + key);
        }
    }

    static class ValidationException extends Exception {
        ValidationException(String message) {
            super(message);
        }
    }

}

File: src/aws/lambdas/incremental_distribution/src/main/java/uk/nhs/nhsx/analyticsevents/PayloadValidator.kt
package uk.nhs.nhsx.analyticsevents

import uk.nhs.nhsx.core.Jackson.readOrNull
import java.io.IOException

class PayloadValidator {
    fun maybeValidPayload(requestBody: String?) = try {
        readOrNull<Map<String, Any>>(requestBody)
            ?.also {
                validateMetadata(it)
                validateEvents(it)
            }
    } catch (e: IOException) {
        null
    } catch (e: ValidationException) {
        null
    } catch (e: IllegalArgumentException) {
        null
    }

    private fun validateMetadata(payload: Map<String, *>) {
        requireMapAttribute(payload, "metadata")
        val metadataRaw = payload["metadata"] as? Map<*, *>
            ?: throw ValidationException("metadata must be a map")
        requireMapAttribute(metadataRaw, "operatingSystemVersion")
        requireMapAttribute(metadataRaw, "latestApplicationVersion")
        requireMapAttribute(metadataRaw, "deviceModel")
        requireMapAttribute(metadataRaw, "postalDistrict")
    }

    private fun validateEvents(payload: Map<String, *>) {
        requireMapAttribute(payload, "events")
        val eventsRaw = payload["events"] as? List<*>
            ?: throw ValidationException("events must be a list")
        for (event in eventsRaw) {
            requireMapAttribute(event, "type")
            requireMapAttribute(event, "version")
            requireMapAttribute(event, "payload")
        }
    }

    private fun requireMapAttribute(mapRaw: Any?, key: String) {
        if (mapRaw !is Map<*, *>) throw ValidationException("expected map type but was " + mapRaw?.javaClass?.simpleName)
        if (!mapRaw.containsKey(key)) throw ValidationException("Map did not contain expected key + $key")
    }

    internal class ValidationException(message: String?) : Exception(message)
}
